lib/train/platforms/uuid.rb

```diff
@@ -0,0 +1,135 @@
+# encoding: utf-8
+
+require 'digest/sha1'
+require 'securerandom'
+require 'json'
+
+module Train::Platforms
+  class UUID
+    def initialize(platform)
+      @platform = platform
+      @backend = @platform.backend
+    end
+
+    def find_or_create_uuid
+      if @platform.unix?
+        unix
+      elsif @platform.windows?
+        windows
+      elsif @platform.name == 'aws'
+        aws
+      elsif @platform.name == 'azure'
+        azure
+      else
+        if @platform[:uuid_command]
+          result = @backend.run_command(@platform[:uuid_command])
+          return uuid_from_string(result.stdout.chomp) if result.exit_status == 0 && !result.stdout.empty?
+        end
+
+        raise 'Could not fine platform uuid! please set a uuid_command for your platform'
+      end
+    end
+
+    private
+
+    def aws
+      # parse the aws account id
+      client = @backend.aws_client(Aws::IAM::Client)
+      arn = client.get_user.user.arn
+      uuid_from_string(arn.split(':')[4])
+    end
+
+    def azure
+      # use subscription_id or tenant_id
+      @backend.options[:subscription_id] || @backend.options[:tenant_id]
+    end
+
+    def windows
+      # check for chef uuid
+      file = @backend.file("#{ENV['SYSTEMDRIVE']}\\chef\\cache\\data_collector_metadata.json")
+      if file.exist? && file.size != 0
+        json = JSON.parse(file.content)
+        return json['node_uuid'] if json['node_uuid']
+      end
+
+      result = @backend.run_command("wmic csproduct get UUID")
+      return result.stdout.split("\r\n")[-1].strip if result.exit_status == 0
+
+      cmd = '(Get-ItemProperty "Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography" -Name "MachineGuid")."MachineGuid"'
+      result = @backend.run_command(cmd)
+      return result.stdout.chomp if result.exit_status == 0
+
+      # see if we have a custom uuid
+      %W(
+        #{ENV['SYSTEMROOT']}\\machine-uuid
+        #{ENV['HOMEDRIVE']}#{ENV['HOMEPATH']}\\.system\\machine-uuid
+      ).each do |path|
+        file = @backend.file(path)
+        return file.content.chomp if file.exist?
+      end
+
+      # cant find anything, try to write a uuid
+      uuid = SecureRandom.uuid
+      result = @backend.run_command("'#{uuid}' | Out-File -encoding ASCII #{ENV['SYSTEMROOT']}\\machine-uuid")
+      if result.exit_status !=0
+        local_uuid_path = "#{ENV['HOMEDRIVE']}#{ENV['HOMEPATH']}\\.system\\machine-uuid"
+        warn "Cannot write uuid to `#{ENV['SYSTEMROOT']}\\machine-uuid`, falling back to #{local_uuid_path} instead"
+        # fallback to user location
+        result = @backend.run_command("new-item -force -path #{local_uuid_path} -value '#{uuid}' -type file")
+        raise "Cannot write uuid to `#{local_uuid_path}\\machine-uuid`" if result.exit_status !=0
+      end
+      uuid
+    end
```

Comment:
I see about 5 methods lurking inside this one. Is that sort of abstraction unconventional in this project?

Reply:
yeah at this point they should be divided up as sub private methods. Thanks for this, ill grab it when I move them over.